torch/functional.py

```diff
@@ -1686,6 +1689,89 @@ def norm(input, p: Optional[Union[float, str]] = "fro", dim=None, keepdim=False,
+
+        shape (int, sequence of ints, or torch.Size): The shape of the arbitrary
+            tensor. All elements must be non-negative.
+
+    Returns:
+        tuple of Tensors: Each ``i``-th tensor in the ouput corresponds with
+        dimension ``i`` of :attr:`shape`. Each tensor has the same shape as
+        ``indices`` and contains one index into dimension ``i`` for each of the
+        flat indices given by ``indices``.
+
+    Example::
+
+        >>> import torch
+        >>> torch.unravel_index(torch.tensor(4), (3, 2))
+        (tensor(2),
+         tensor(0))
+
+        >>> torch.unravel_index(torch.tensor([4, 1]), (3, 2))
+        (tensor([2, 0]),
+         tensor([0, 1]))
+
+        >>> torch.unravel_index(torch.tensor([0, 1, 2, 3, 4, 5]), (3, 2))
+        (tensor([0, 0, 1, 1, 2, 2]),
+         tensor([0, 1, 0, 1, 0, 1]))
+
+        >>> torch.unravel_index(torch.tensor([1234, 5678]), (10, 10, 10, 10))
+        (tensor([1, 5]),
+         tensor([2, 6]),
+         tensor([3, 7]),
+         tensor([4, 8]))
+
+        >>> torch.unravel_index(torch.tensor([[1234], [5678]]), (10, 10, 10, 10))
+        (tensor([[1], [5]]),
+         tensor([[2], [6]]),
+         tensor([[3], [7]]),
+         tensor([[4], [8]]))
+
+        >>> torch.unravel_index(torch.tensor([[1234], [5678]]), (100, 100))
+        (tensor([[12], [56]]),
+         tensor([[34], [78]]))
+    """
+    if has_torch_function_unary(indices):
+        return handle_torch_function(
+            unravel_index, (indices,), indices, shape=shape)
+    res_tensor = _unravel_index(indices, shape)
+    return res_tensor.unbind(-1)
+
+def _unravel_index(indices: Tensor, shape: Union[int, Sequence[int], torch.Size]) -> Tensor:
+    torch._check_type(
+        not indices.is_complex() and not indices.is_floating_point() and not indices.dtype == torch.bool,
+        lambda: f"expected 'indices' to be integer dtype, but got {indices.dtype}")
+
+    torch._check_type(
+        isinstance(shape, (int, Sequence, torch.Size)),
+        lambda: f"expected 'shape' to be int, sequence of ints, or torch.Size, but got {type(shape)}")
+
+    if isinstance(shape, int):
+        shape = torch.Size([shape])
+    elif isinstance(shape, Sequence):
+        for dim in shape:
+            torch._check_type(
+                isinstance(dim, int),
+                lambda: f"expected 'shape' sequence to only contain ints, but got {type(dim)}")
+        shape = torch.Size(shape)
+
+    torch._check_value(
+        all(dim >= 0 for dim in shape),
+        lambda: f"'shape' cannot have negative values, but got {tuple(shape)}")
+
+    coefs = list(reversed(list(itertools.accumulate(reversed(shape[1:] + torch.Size([1])), func=operator.mul))))
+    res = indices.unsqueeze(-1).div(
+        torch.tensor(coefs, device=indices.device, dtype=torch.int64),
+        rounding_mode='trunc') % torch.tensor(shape, device=indices.device, dtype=torch.int64)
```

Comment:
Isn't it more efficient to do the cumprod as a tensor already?

```python
shape = torch.tensor((*shape, 1), device=indices.device)
coefs = shape[1:].flip(0).cumprod(0).flip(0)
res = torch.div(indices.unsqueeze(-1), coefs, rounding_mode='trunc') % shape[:-1]
```

Also, I believe `torch.tensor` will automatically cast Python integers to long (`torch.int64`).

Reply:
see https://github.com/pytorch/pytorch/pull/110580#discussion_r1350987816

About the second comment, as they say, explicit is better than implicit :)